### ISSUE TITLE
Contributions

### ISSUE THREAD

**Issue by jvskriubakken** [cfnz/muirwik#9]
Hi cfnz! I find no other way to contact you than creating an issue on your muirwik project. I might be able to help you in making this Kotlin wrapper for Material UI. I'm currently using Javascript (create-react-app) and Material-UI at work, but are looking for opportunities to possibly convert to using Kotlin. I've from before create frontend apps with Kotlin from ground up. But I believe that the be able to succeed with Kotlin JS, there is a need for wrappers the most useful UI libraries, like Material UI, Highcharts, Mapbox, etc.

How is going with mapping mui to Kotlin? Any challenges?

Is muirwik deployed as a NPM dependency?

Regards,

Jørund

**Comment by cfnz** [comment]
Hi Jørund,

Thanks for making contact. Sorry for the late reply, holidays etc, have kept me away/busy :-)

> How is going with mapping mui to Kotlin? Any challenges?

Yip, I have had a few challenges... spent lots of hours on them, and I am sure if I was more familiar with Javascript/Node/Webpack, they would have been solved a lot quicker! But it has been a good leaning experience, and currently I don't have any brick walls that I am banging my head against, I am sure things could be done better, but it is all working.

The latest thing I have done (and not committed yet), is made the component library a maven dependency. Currently I just deploy to maven local, but should look at making a public repository for that. This makes the projects using the library compile much quicker (and probably easier to get going too) as they don't recompile the components each time. (I thought I already had this working with the Gradle projects in the current repository, but I think this still compiles the component project). When I get a bit of time, I will commit this.

> Is muirwik deployed as a NPM dependency?

No, but I don't think it needs to be (but could be wrong :-) ), especially once I create the Gradle/Maven public repo as mentioned above, all the files should get pulled down from that in a way that Kotlin can use including the compiled JS files. Any other JS files will get created by Kotlin.

There are still lots of things to do... Material UI is quite a fast moving project, so I am having trouble just keeping up with that, and I have not mapped every component there yet (but most are done and working).

Happy for any help/suggestions/advice where needed :-)

Regards
Colin

**Comment by jvskriubakken** [comment]
Great to hear from you Colin!

Yeah! I think you have achieved something great considering that you have jumped on this with no little experience with both Kotlin and the Node/Webpack world. My experience with Node and Webpack is pretty low as I until now have relied on using create-react-app which gives zero configuration of webpack for react apps. I guess you have seen Jetbrains similar create-react-kotlin-app:
https://github.com/JetBrains/create-react-kotlin-app

I believe that this project, muirwik will attract more users if it was published both as a maven dependency and a npm package. Because, if you are using Kotlin JS through create-react-kotlin-app, then you are limited to npm packages (as far as I know). Also, I've heard that Parcel is working on Kotlin support in their next large version, so if I guess people are limited to NPM packages there too.

Did you use the ts2kt converter?: https://github.com/Kotlin/ts2kt
Or did you do all the mappings manually?

Cheers,

Jørund

**Comment by cfnz** [comment]
Hi Jørund,

Yes, I have seen create-react-kotlin-app, and played around with it a bit (it is how I got started). I hit a few issues (can't recall now what they were - possibly trying to use one kotlin project from the other), but I remember "ejecting" the create-react-kotlin-app and examining the result and getting a bit further.

There is also a Gradle plugin that helps with the JS side of things, but I also found [this video](https://www.youtube.com/watch?v=1Pu0TYJJ2Tw), and somewhere in there the author talks about using Gradle, and if you use Gradle you may be better off handling the JS/Node/Webpack stuff yourself, so that is what I tried and it seemed to work the best. I think once you get going you will want to be managing dependencies for the Kotlin as well as the JS side of things, and I don't know if the NPM side of things will handle both, but Gradle does (well, what the the guy in the video did, and I followed, was to use NPM and Webpack from within Gradle... so you still have a package and webpack config).

You mention Parcel, I have not really had a look at that at all... might check it out in the next few days if I get the chance...

In terms of ts2kt, yes, I used that but only for a few files, the rest were done manually as ts2kt did not really handle them very well.

Regards
Colin

**Comment by galex** [comment]
Hello! 
Do you have some news about this? I'd really like to use this library instead of re-inventing all what you did here, which is super impressive btw. 

I barely understand what is going on there, as I am not familiar with the React, webpack but coming from Android I know enough about Gradle to have deployed a few libraries to bintray/jcenter.

I would love to help as well if I can.

**Comment by cfnz** [comment]
Hi, thanks for that.

I hope to have some time in the next few days (except for today) to spend some time on it.

Had a quick look last night at bintray/jcenter and setup an account... If I get stuck on configuring/deployment I shall reach out for some help.

Regards
Colin

**Comment by galex** [comment]
Cool! I guess the easiest is to look at the official kotlin wrappers build.gradle, [like for example this one](https://github.com/cfnz/kotlin-wrappers/blob/7e88025bdace48305cc9b3e64e5097efd39eab95/build.gradle#L53).

**Comment by galex** [comment]
Hey,

The gradle wrapper goal is to have it embedded in the project so users don't have to download it on their machine themselves. It is generally recommended to use it and considered a good practice to run every task through it, example `./gradlew clean build`

For the structure of the project for a sample/starterapp, it is up to whatever makes more sense to you, but I tend to put a sample app within the git repository of the library, so the file structure would look like this:
- lib (folder with its own build.gradle + src, etc)
- sample (folder with its own build.gradle, src, etc)
- settings.gradle (with include "lib", "sample")

So say you would build and deploy the lib, you would use `./gradlew :lib:publishArtifactToX`. The users downloading the project will run the sample app like this: `./gradlew :sample:run`. But if it makes more sense to you to have two different git repositories, it is fine as well. 

I wouldn't recommend a git submodule in this case as you are distributing publicly the lib via jCenter anyway. 

Hope this helps!

**Comment by cfnz** [comment]
Thanks for that advice.

I quite liked having the starter-app under the main project too... and this is how I started. But I have now taken the "But if it makes more sense to you" approach in that I was keen to have the users of the project get an easy start without having to bother with code they don't need... so I have ended up creating a new project [muirwik-starterapp](https://github.com/cfnz/muirwik-starterapp). Hopefully it works off the bat, tried it with a VM and needed a few tweaks, but think it works now...

I have created a new release that uploads to bintray, just waiting for jcenter. I also made breaking changes to package names and modules... (if anyone is using this code, I can give the search and replace strings... it is not a major).

Regards
Colin
